### PR TITLE
fix(amplify-codegen-appsync-model-plugin): pass targetName for hasOne relationships

### PR DESCRIPTION
**Pull request by lawmicha**
*Issue #, if available:*
https://github.com/aws-amplify/amplify-ios/issues/920

*Description of changes:*
This adds the targetName for hasOne relationships, closely mirroring the same logic applied for belongsTo. 

*Testing done*
schema
```
type Project2 @model {
  id: ID!
  name: String
  teamID: ID! // Explicit field for Team association
  team: Team2 @connection(fields: ["teamID"])
}

type Team2 @model {
  id: ID!
  name: String!
}
```

Generated swift schema class without this change:
```swift
// swiftlint:disable all
import Amplify
import Foundation

extension Project2 {
  // MARK: - CodingKeys 
   public enum CodingKeys: String, ModelKey {
    case id
    case name
    case teamID
    case team
  }
  
  public static let keys = CodingKeys.self
  //  MARK: - ModelSchema 
  
  public static let schema = defineSchema { model in
    let project2 = Project2.keys
    
    model.pluralName = "Project2s"
    
    model.fields(
      .id(),
      .field(project2.name, is: .optional, ofType: .string),
      .field(project2.teamID, is: .required, ofType: .string),
      .hasOne(project2.team, is: .optional, ofType: Team2.self, associatedWith: Team2.keys.id)
    )
    }
}
```
Generated schema with this change:
```swift
.hasOne(project2.team, is: .optional, ofType: Team2.self, associatedWith: Team2.keys.id, targetName: "teamID")
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by ammarkarachi** [comment]
@AaronZyLee Can you review?

**Review by AaronZyLee — COMMENTED** [review]
LGTM but I am still investigating and looking out for a general solution for other SDKs. And what I am concerned for this solution is that `teamID` is not optional when you create a `Project`, not only in modelgen but also in `CreateProjectInput` in appsync. When refering to A has one B, it means that A does not necessarily need B but B is dependent on A. If we change the `teamID` to an optional field, it will throw the following error by transfomer:

```
InvalidDirectiveError: All fields provided to an @connection must be non-null scalar or enum fields.
```
The reason is that when referring to an identifier in `@connection(fields:[])` the param should not be nullable.


On the other hand, I am not sure our [document](https://docs.amplify.aws/cli/graphql-transformer/connection#has-one) has a good guide on how to design a one-to-one schema. Why do we need an identifier in a `HAS_ONE` relation? Do we have any scenario that using `project.teamID` is better than using `project.team.id`? 

In the `HAS_MANY` and `BELONGS_TO` bi-way connection(one to many), however, it is really necessary to use an identifier. Say post has many comments, and in comment schema we add a `postID` as well as a @key to define `postID` as a GSI, from which the post can have a query to get all comments for `post.comments`.

In conclusion, I think a better approach is that we do not use identifier, aka `teamID` from the example schema, in a one to one relation. Overall, the following schema should be an expected one-to-one schema in our guide, which causes less trouble. And one thing to mention is that the following pattern is also used in the Amplify Admin UI.

```GraphQL
type Project @model {
  id: ID!
  name: String
  team: Team @connection
}

type Team @model {
  id: ID!
  name: String!
}
```

**Review by AaronZyLee — APPROVED** [review]

**Comment by lawmicha** [comment]
@AaronZyLee  Thanks for reviewing, will loop back here to continue the conversation regarding your comment. Can we merge this in first? I believe your approval doesn't have write access yet. @attilah or someone with write access required can you also approve to merge this in?

**Comment by lawmicha** [comment]
> LGTM but I am still investigating and looking out for a general solution for other SDKs. And what I am concerned for this solution is that `teamID` is not optional when you create a `Project`, not only in modelgen but also in `CreateProjectInput` in appsync. When refering to A has one B, it means that A does not necessarily need B but B is dependent on A. If we change the `teamID` to an optional field, it will throw the following error by transfomer:
> 
> ```
> InvalidDirectiveError: All fields provided to an @connection must be non-null scalar or enum fields.
> ```
> 
> The reason is that when referring to an identifier in `@connection(fields:[])` the param should not be nullable.
> 

I also noticed this behavior and believe a non optional team ID is correct. A team can exist without a project, but a project cannot exist without a team. you should never have orphaned projects, and always associated with a team. If you want to have a orphaned project, you may as the developer create a dummy placeholder team to associate your orphaned projects.

> On the other hand, I am not sure our [document](https://docs.amplify.aws/cli/graphql-transformer/connection#has-one) has a good guide on how to design a one-to-one schema. Why do we need an identifier in a `HAS_ONE` relation? Do we have any scenario that using `project.teamID` is better than using `project.team.id`?
> 
> In the `HAS_MANY` and `BELONGS_TO` bi-way connection(one to many), however, it is really necessary to use an identifier. Say post has many comments, and in comment schema we add a `postID` as well as a @key to define `postID` as a GSI, from which the post can have a query to get all comments for `post.comments`.
> 
> In conclusion, I think a better approach is that we do not use identifier, aka `teamID` from the example schema, in a one to one relation. Overall, the following schema should be an expected one-to-one schema in our guide, which causes less trouble. And one thing to mention is that the following pattern is also used in the Amplify Admin UI.
> 
> ```graphql
> type Project @model {
>   id: ID!
>   name: String
>   team: Team @connection
> }
> 
> type Team @model {
>   id: ID!
>   name: String!
> }
> ```

I believe we need to support both explicit fields using the key directive and fields parameter in connections and implicit fields using the older connection directive (without the `name` parameter/omitting all parameters like u have above). In your above example. But how I understood the documentation was that explicit field with key directive and fields is the newer version of the connection directive usage and is the recommended way

**Comment by ammarkarachi** [comment]
LGTM

**Review by ammarkarachi — APPROVED** [review]

**Review by SwaySway — APPROVED** [review]
